statistics/builder.go

```diff
@@ -184,3 +187,188 @@ func BuildColumnHist(ctx sessionctx.Context, numBuckets, id int64, collector *Sa
+		// TopN includes all sample data
+		return hg, topn, nil
+	}
+	// Since bucket count is increased by sampleFactor, so the actual max values per bucket is
+	// floor(valuesPerBucket/sampleFactor)*sampleFactor, which may less than valuesPerBucket,
+	// thus we need to add a sampleFactor to avoid building too many buckets.
+	valuesPerBucket := float64(count)/float64(numBuckets) + sampleFactor
+	ndvFactor := float64(uint64(count)-topn.TotalCount()) / float64(int(hg.NDV)-len(topn.TopN))
+	if ndvFactor > sampleFactor {
+		ndvFactor = sampleFactor
+	}
+	bucketIdx := 0
+	var lastCount int64
+
+	// Step3: build histogram with the rest samples
+	if len(samples) > 0 {
+		hg.AppendBucket(&samples[0].Value, &samples[0].Value, int64(sampleFactor), int64(ndvFactor))
+	}
+	for i := int64(1); i < int64(len(samples)); i++ {
+		cmp, err := hg.GetUpper(bucketIdx).CompareDatum(sc, &samples[i].Value)
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
+		totalCount := float64(i+1) * sampleFactor
+		if cmp == 0 {
+			// The new item has the same value as current bucket value, to ensure that
+			// a same value only stored in a single bucket, we do not increase bucketIdx even if it exceeds
+			// valuesPerBucket.
+			hg.Buckets[bucketIdx].Count = int64(totalCount)
+			if hg.Buckets[bucketIdx].Repeat == int64(ndvFactor) {
+				hg.Buckets[bucketIdx].Repeat = int64(2 * sampleFactor)
+			} else {
+				hg.Buckets[bucketIdx].Repeat += int64(sampleFactor)
+			}
+		} else if totalCount-float64(lastCount) <= valuesPerBucket {
+			// The bucket still have room to store a new item, update the bucket.
+			hg.updateLastBucket(&samples[i].Value, int64(totalCount), int64(ndvFactor))
+		} else {
+			lastCount = hg.Buckets[bucketIdx].Count
+			// The bucket is full, store the item in the next bucket.
+			bucketIdx++
+			hg.AppendBucket(&samples[i].Value, &samples[i].Value, int64(totalCount), int64(ndvFactor))
+		}
+	}
+
+	// Compute column order correlation with handle.
+	if sampleNum == 1 {
+		hg.Correlation = 1
+		return hg, topn, nil
+	}
+	// X means the ordinal of the item in original sequence, Y means the ordinal of the item in the
+	// sorted sequence, we know that X and Y value sets are both:
+	// 0, 1, ..., sampleNum-1
+	// we can simply compute sum(X) = sum(Y) =
+	//    (sampleNum-1)*sampleNum / 2
+	// and sum(X^2) = sum(Y^2) =
+	//    (sampleNum-1)*sampleNum*(2*sampleNum-1) / 6
+	// We use "Pearson correlation coefficient" to compute the order correlation of columns,
+	// the formula is based on https://en.wikipedia.org/wiki/Pearson_correlation_coefficient.
+	// Note that (itemsCount*corrX2Sum - corrXSum*corrXSum) would never be zero when sampleNum is larger than 1.
+	itemsCount := float64(sampleNum)
+	corrXSum := (itemsCount - 1) * itemsCount / 2.0
+	corrX2Sum := (itemsCount - 1) * itemsCount * (2*itemsCount - 1) / 6.0
+	hg.Correlation = (itemsCount*corrXYSum - corrXSum*corrXSum) / (itemsCount*corrX2Sum - corrXSum*corrXSum)
+	return hg, topn, nil
+}
```

Comment:
This part is duplicated with `BuildColumnHist`, so please create a separate function for it.